codegen/apt/src/main/java/io/helidon/codegen/apt/AptContextImpl.java

```diff
@@ -0,0 +1,160 @@
+                                  new AptLogger(env, options),
+                                  scope,
+                                  module.orElse(null));
+    }
+
+    @Override
+    public ProcessingEnvironment aptEnv() {
+        return env;
+    }
+
+    @Override
+    public Optional<TypeInfo> typeInfo(TypeName typeName) {
+        return AptTypeInfoFactory.create(this, typeName);
+    }
+
+    @Override
+    public Optional<TypeInfo> typeInfo(TypeName typeName, Predicate<TypedElementInfo> elementPredicate) {
+        return AptTypeInfoFactory.create(this, typeName, elementPredicate);
+    }
+
+    @Override
+    public Optional<ModuleInfo> module() {
+        return Optional.ofNullable(moduleInfo);
+    }
+
+    private static Optional<ModuleInfo> findModule(Filer filer) {
+        // expected is source location
+        try {
+            FileObject resource = filer.getResource(StandardLocation.SOURCE_PATH, "", "module-info.java");
+            try (InputStream in = resource.openInputStream()) {
+                return Optional.of(ModuleInfoSourceParser.parse(in));
+            }
+        } catch (IOException ignored) {
+            // it is not in sources, let's see if it got generated
+        }
+        // generated
+        try {
+            FileObject resource = filer.getResource(StandardLocation.SOURCE_OUTPUT, "", "module-info.java");
+            try (InputStream in = resource.openInputStream()) {
+                return Optional.of(ModuleInfoSourceParser.parse(in));
+            }
+        } catch (IOException ignored) {
+            // not in generated source either
+        }
+        // we do not see a module info
+        return Optional.empty();
+    }
+
+    private static CodegenScope guessScope(ProcessingEnvironment env, CodegenOptions options) {
+        CodegenScope scopeFromOptions = CodegenOptions.CODEGEN_SCOPE.findValue(options).orElse(null);
+
+        if (scopeFromOptions != null) {
+            return scopeFromOptions;
+        }
+        try {
+            URI resourceUri = env.getFiler()
+                    .getResource(StandardLocation.CLASS_OUTPUT, "does.not.exist", "DefinitelyDoesNotExist")
+                    .toUri();
+
+            // should be something like:
+            // file:///projects/helidon_4/inject/tests/resources-inject/target/test-classes/does/not/exist/DefinitlyDoesNotExist
+            String resourceUriString = resourceUri.toString();
+            if (!resourceUriString.endsWith("/does/not/exist/DefinitelyDoesNotExist")) {
+                // cannot guess, not ending in expected string, assume production scope
+                return CodegenScope.PRODUCTION;
+            }
+            // full URI
+            resourceUriString = resourceUriString
+                    .substring(0, resourceUriString.length() - "/does/not/exist/DefinitelyDoesNotExist".length());
+            // file:///projects/helidon_4/inject/tests/resources-inject/target/test-classes
```

Comment:
Project relative reference would be better

Reply:
It is a project relative reference:
`.getResource(StandardLocation.CLASS_OUTPUT, "does.not.exist", "DefinitelyDoesNotExist")`
The leading slash is because we are looking at path before the suffix of the package and class name.